<a href="https://colab.research.google.com/github/luciekash/WebApp/blob/master/Lucy_mwenda_intro_to_machine_learn.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Instructions**

Background Information
HR analytics is revolutionising the way human resources departments operate, leading
to higher efficiency and better results overall. Human resources have been using
analytics for years. However, the collection, processing, and analysis of data have been
largely manual, and given the nature of human resources dynamics and HR KPIs, the
approach has been constraining HR. Therefore, it is surprising that HR departments
woke up to the utility of machine learning so late in the game.


**Pre-requisites**

In [3]:
#import libraries
import pandas as pd
#import decision tree from the sklearn library
from sklearn.tree import DecisionTreeClassifier

# load and preview dataset
hr_train = pd.read_csv ('https://bit.ly/2ODZvLCHRDataset')
hr_train


employee_id         department  ... avg_training_score is_promoted
0            65438  Sales & Marketing  ...                 49           0
1            65141         Operations  ...                 60           0
2             7513  Sales & Marketing  ...                 50           0
3             2542  Sales & Marketing  ...                 50           0
4            48945         Technology  ...                 73           0
...            ...                ...  ...                ...         ...
54803         3030         Technology  ...                 78           0
54804        74592         Operations  ...                 56           0
54805        13918          Analytics  ...                 79           0
54806        13614  Sales & Marketing  ...                 45           0
54807        51526                 HR  ...                 49           0

[54808 rows x 14 columns]

In [18]:
hr_train['is_promoted'].value_counts()

0    47853
1     4546
Name: is_promoted, dtype: int64

In [6]:
# preview dataset shape
hr_train.shape

(54808, 14)

In [5]:
# preview datatypes
hr_train.dtypes

employee_id               int64
department               object
region                   object
education                object
gender                   object
recruitment_channel      object
no_of_trainings           int64
age                       int64
previous_year_rating    float64
length_of_service         int64
KPIs_met >80%             int64
awards_won?               int64
avg_training_score        int64
is_promoted               int64
dtype: object

In [16]:
#Identify duplicates records in the data
hr_train.duplicated().sum()


0

In [17]:
# look for missing records
hr_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 52399 entries, 0 to 54807
Data columns (total 14 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   employee_id           52399 non-null  int64  
 1   department            52399 non-null  object 
 2   region                52399 non-null  object 
 3   education             52399 non-null  object 
 4   gender                52399 non-null  object 
 5   recruitment_channel   52399 non-null  object 
 6   no_of_trainings       52399 non-null  int64  
 7   age                   52399 non-null  int64  
 8   previous_year_rating  52399 non-null  float64
 9   length_of_service     52399 non-null  int64  
 10  KPIs_met >80%         52399 non-null  int64  
 11  awards_won?           52399 non-null  int64  
 12  avg_training_score    52399 non-null  int64  
 13  is_promoted           52399 non-null  int64  
dtypes: float64(1), int64(8), object(5)
memory usage: 6.0+ MB


In [9]:
hr_train.isnull().sum()

employee_id                0
department                 0
region                     0
education               2409
gender                     0
recruitment_channel        0
no_of_trainings            0
age                        0
previous_year_rating    4124
length_of_service          0
KPIs_met >80%              0
awards_won?                0
avg_training_score         0
is_promoted                0
dtype: int64

From the above analysis we see education has 2409 and previous_year_rating has 4124 missing values.

In [20]:
#drop null values and columns that have nno useful information
hr_train.drop(['employee_id'], axis=1, inplace=True)

hr_train.dropna(subset=['education'],inplace=True)

hr_train['previous_year_rating'].fillna(value=hr_train['previous_year_rating'].mean(), inplace=True)


In [21]:
hr_train.isna().sum()

department              0
region                  0
education               0
gender                  0
recruitment_channel     0
no_of_trainings         0
age                     0
previous_year_rating    0
length_of_service       0
KPIs_met >80%           0
awards_won?             0
avg_training_score      0
is_promoted             0
dtype: int64

In [22]:
hr_train.head()

department     region  ... avg_training_score is_promoted
0  Sales & Marketing   region_7  ...                 49           0
1         Operations  region_22  ...                 60           0
2  Sales & Marketing  region_19  ...                 50           0
3  Sales & Marketing  region_23  ...                 50           0
4         Technology  region_26  ...                 73           0

[5 rows x 13 columns]

**Data Modeling**

In [29]:
hr_train=pd.get_dummies(hr_train)
hr_train.head()

no_of_trainings  ...  recruitment_channel_sourcing
0                1  ...                             1
1                1  ...                             0
2                1  ...                             1
3                2  ...                             0
4                1  ...                             0

[5 rows x 59 columns]

In [32]:
#we first start by creating two variable to store our dataset

features = hr_train.drop(['is_promoted'], axis=1)
target = hr_train['is_promoted']

print(features.shape)
print(target.shape)


(52399, 58)
(52399,)


In [34]:
# create an empty model and assign it to a variable
model = DecisionTreeClassifier()
model


DecisionTreeClassifier()

In [35]:
# train a model by calling the fit() method
model.fit(features, target)

DecisionTreeClassifier()

In [38]:
#making  predictions using a model.
hr_test = pd.read_csv("/content/test_2umaH9m.csv")
hr_test

employee_id         department  ... awards_won? avg_training_score
0             8724         Technology  ...           0                 77
1            74430                 HR  ...           0                 51
2            72255  Sales & Marketing  ...           0                 47
3            38562        Procurement  ...           0                 65
4            64486            Finance  ...           0                 61
...            ...                ...  ...         ...                ...
23485        53478              Legal  ...           0                 61
23486        25600         Technology  ...           0                 74
23487        45409                 HR  ...           0                 50
23488         1186        Procurement  ...           0                 70
23489         5973         Technology  ...           0                 89

[23490 rows x 13 columns]

In [39]:
#drop null values and columns that have nno useful information
hr_test.drop(['employee_id'], axis=1, inplace=True)

hr_test.dropna(subset=['education'],inplace=True)

hr_test['previous_year_rating'].fillna(value=hr_test['previous_year_rating'].mean(), inplace=True)

In [40]:
hr_test=pd.get_dummies(hr_test)
hr_test.head()

no_of_trainings  ...  recruitment_channel_sourcing
0                1  ...                             1
1                1  ...                             0
2                1  ...                             0
3                3  ...                             0
4                1  ...                             1

[5 rows x 58 columns]

In [41]:
new_features = hr_test

new_features['is_promoted']= model.predict(new_features)

new_features.head()

no_of_trainings  age  ...  recruitment_channel_sourcing  is_promoted
0                1   24  ...                             1            0
1                1   31  ...                             0            0
2                1   31  ...                             0            0
3                3   31  ...                             0            0
4                1   30  ...                             1            0

[5 rows x 59 columns]

In [42]:
new_features['is_promoted'].value_counts()

0    20393
1     2063
Name: is_promoted, dtype: int64